src/Features/Core/Portable/CodeLens/CodeLensFindReferenceProgress.cs

```diff
@@ -0,0 +1,173 @@
+    /// </remarks>
+    internal sealed class CodeLensFindReferencesProgress : IFindReferencesProgress, IDisposable
+    {
+        /// <summary>
+        /// this token is linked to an aggregate token that is passed to the Find References
+        /// operation, this is used solely to trigger the cancellation of an in progress
+        /// find references operation, when the references count hits a given cap.
+        /// </summary>
+        private readonly CancellationTokenSource _cancellationTokenSource;
+        private readonly CancellationTokenSource _aggregateCancellationTokenSource;
+        private readonly SyntaxNode _queriedNode;
+        private readonly ISymbol _queriedSymbol;
+        private readonly ConcurrentSet<Location> _locations;
+
+        /// <remarks>
+        /// If the cap is 0, then there is no cap.
+        /// </remarks>
+        public int SearchCap { get; }
+
+        /// <summary>
+        /// The cancellation token that aggregates the original cancellation token + this progress
+        /// </summary>
+        public CancellationToken CancellationToken => _aggregateCancellationTokenSource.Token;
+
+        public bool SearchCapReached => SearchCap != 0 && ReferencesCount > SearchCap;
+
+        public int ReferencesCount => _locations.Count;
+
+        public ImmutableArray<Location> Locations => _locations.ToImmutableArray();
+
+        public CodeLensFindReferencesProgress(
+            ISymbol queriedDefinition,
+            SyntaxNode queriedNode,
+            int searchCap,
+            CancellationToken cancellationToken)
+        {
+            _queriedSymbol = queriedDefinition;
+            _queriedNode = queriedNode;
+            _cancellationTokenSource = new CancellationTokenSource();
+            _aggregateCancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(
+                _cancellationTokenSource.Token, cancellationToken);
+            _locations = new ConcurrentSet<Location>();
+
+            SearchCap = searchCap;
+        }
+
+        public void OnCompleted()
+        {
+        }
+
+        /// <summary>
+        /// Returns partial symbol locations whose node does not match the given syntaxNode
+        /// </summary>
+        /// <param name="symbol">Symbol whose locations are queried</param>
+        /// <param name="syntaxNode">Syntax node to compare against to exclude location - actual location being queried</param>
+        /// <param name="cancellationToken">Cancellation token</param>
+        /// <returns>Partial locations</returns>
+        private static IEnumerable<Location> GetPartialLocations(ISymbol symbol, SyntaxNode syntaxNode, CancellationToken cancellationToken)
+        {
+            // Returns nodes from source not equal to actual location and matches the definition locations to syntax references, ignores metadata locations
+            return from syntaxReference in symbol.DeclaringSyntaxReferences
```

Comment:
probably micro optimization, but since Linq is expansive, and feels like this could be called a lot. and common case probably will be when symbol.DeclaringSyntaxReferences and symbol.Locations == 1.

how about adding special case for those? to reduce CPU and allocations?